Split out foundation

imduffy15 opened the pull request:
This pull requests adds an extra step between transitioning from foundation to base.

In the event we're using multiple different hypervisors or cloud providers it would be nice to use a single vmdk created from an ISO installation on these several platforms without rebuilding the disk for each.

The foundation image exists to create a JeOS VMDK. The Membrane exists to layer on-top platform specific tools: vmware tools, virtual box guest additions, cloud-init, walinuxagent etc.

Additionally, i've modified foundation to output a json object that can be used as an input in the membrane box. Using jq we can do a merge of the two variable files:

```
cat <half-filled-variables> <dynamic-variables> | jq add -s -r > variables.json
packer build -var-file variables.json packer.json
``
```

samdunne commented on boxes/foundation/www/ks.cfg:
```diff
@@ -0,0 +1,327 @@
+cpio
+cracklib
+cracklib-dicts
+cronie
```
Do we want this?

imduffy15 replied:
Yes. Its part of crontab.

samdunne commented on scripts/aws.sh:
```diff
@@ -0,0 +1,79 @@
+ - package-update-upgrade-install
+ - puppet
+ - chef
+ - salt-minion
```
Why salt and mcollective?

imduffy15 replied:
Its part of the cloud-init default config.

samdunne replied:
Neat

samdunne commented on boxes/membrane/packer.json:
```diff
@@ -4,10 +4,10 @@
       "headless": "{{ user `headless` }}",
       "name": "{{ user `virtualbox_identifier` }}",
       "output_directory": "builds/{{ user `group_id` }}.{{ user `virtualbox_identifier` }}.{{ user `artifact_id` }}/{{ user `major_version` }}.{{ user `minor_version` }}.{{ user `patch_version` }}.{{ user `incremental_version` }}/",
+      "shutdown_command": "/sbin/halt -h -p",
```
Lexicographical order :(

samdunne commented on boxes/membrane/variables.json:
```diff
@@ -0,0 +1,14 @@
+  "major_version": "0",
+  "minor_version": "0",
+  "patch_version": "0",
+  "virtualbox_source_virtualdisk_path": "builds/com.gauchoconsulting.cloud.virtualbox.centos6-foundation/0.0.0.0/com.gauchoconsulting.cloud.virtualbox.centos6-foundation-0.0.0.0-disk1.vmdk",
```
Shouldn't this be `builds/com/gauchoconsulting...`?

imduffy15 committed:
Split foundation out

This PR splits foundation out into a mixture of the `foundation` image and a `membrane` image.

The foundation image is focused on just enough OS (JeOS), it effectively converts the installation
medium into a bootable VMDK.

The membrane image is for layer ontop specific hypervisor/cloud-provider requirements for example:

 - VirtualBox = VirtualBox Guest Additions
 - Vagrant = A vagrant user and the vagrant SSH key
 - Amazon Web Services = cloud-init

Signed-off-by: Ian Duffy <ian@ianduffy.ie>
